### DISCUSSION TITLE
如何搭建HTTPS代理？

### DISCUSSION THREAD

**Discussion by aaro-n** [v2fly/v2ray-core#1515] [Q&A]
### 本人情况说明：
自建已经使用v2ray搭建一个代理，并且使用良好（感谢v2ray项目）。目前碰到一些问题：
1. 使用场景
    1.1 手机上有些软件需要使用v2ray才能访问，软件本身支持设置代理，例如Telegram（socks，MTProxy），Twidere（https）。我想让软件直接设置https代理，不用开v2ray全局（我知道可以设置哪些软件不走代理，但就是不想运行个科学上网软件）。
     1.2 电脑上chrome的SwitchyOmega插件可以设置https代理，桌面端Spotify也可以使用https代理......,电脑上运行的v2ray转socks或https，也可以正常使用。我想让翻墙无感化。
2. 一些想法
     2.1 由于有vps，想要在vps搭建http代理，最好可以在前边使用nginx作为前置。用户-->nginx前置代理-->http翻墙代理，因为还有其他项目，无法空出80和443端口。
      2.2 一些疑虑 因为想要软件直接使用，软件的代理界面会出现用户名和密码，在使用https的情况下，用户名和密码会不会泄漏？例子：**https://user:pass@github.com** ，这个**user:pass**会不会在公网被明文传输？（因为会用到nginx的htpasswd用户认证，不知道理解的对不对）。
       2.3 GFW HTTP翻墙使用量不会太大，并且有被墙IP或域名的心里准备，被封也无所无，但也希望不要被墙，有些想法：如果**user:pass**不被明文传输，有没有那种**user:pass**正确，就可以正常翻墙，无**user:pass**或者**user:pass**错误就展示一个正常网站，加强隐秘性。
3. 查找资料
     因为使用v2ray非常好，查找v2ray资料，发现v2ray的http代理不符合我的要求（不想要翻墙客户端打开），用中文和英文在Google上搜索了，都没有找到符合要求的项目（可能关键词不对），因此来这里求助。
   ### 总结：因为一些软件内置http代理，想要在不开翻墙代理的情况下，直接填入域名和用户名密码直接使用。
**一些软件代理设置界面截图：**
<img width="910" alt="截屏2022-01-03 下午4 24 51" src="https://user-images.githubusercontent.com/45537680/147911353-24f2cb63-a874-43fa-9245-fae93a463db4.png">
<img width="555" alt="截屏2022-01-03 下午4 26 06" src="https://user-images.githubusercontent.com/45537680/147911401-8f354fd0-8c8e-4516-b6fb-8acc83cbb060.png">
<img width="551" alt="截屏2022-01-03 下午4 28 28" src="https://user-images.githubusercontent.com/45537680/147911423-6fed1457-4351-406c-a530-72ff6bfb6bd8.png">

**Comment by rowheel** [comment]
This shall not be achieved with v2ray-core; try caddy-server instead.

**Comment by sundi117** [comment]
楼主这个问题想出方案了吗？

**Comment by mydogshitgold** [comment]
HTTPS is TLS coupled with HTTP in such a way that once the TCP connection is established, a TLS handshake is performed, creating an encrypted tunnel between the client and the server. All HTTP traffic within the tunnel is encrypted, including the URL (so your `user:pass` is encrypted). However, the TLS handshake does expose information such as the ports, addresses, application layer protocol (alpn), domain name and server certificate.

Theoretically, you should be able to use a web server as a TLS termination proxy for the V2Ray http proxy and simultaneously serve web pages on the same port. Refer to your web server documentation to learn how to do this.

However, HTTP proxy over TLS (aka HTTPS proxy) has many disadvantages.

- It provides weak obfuscation, which can lead to blocking.
- It lacks multiplexing.
- Communication over an HTTPS proxy suffers from much higher latency than stateless protocols such as VMESS, Shadowsocks. Each request sent through an HTTPS proxy requires at least 2 additional round trips (TCP, TLS).

EDIT: less confusing terminology

**Reply by nursery01** [reply]
> However, the TLS handshake does expose information such as the 5 tuple, domain name and server certificate.

what is 5 tuple?

**Reply by mydogshitgold** [reply]
> what is 5 tuple?

edited

**Reply by nursery01** [reply]
> However, the TLS handshake does expose information such as the ports, addresses, application layer protocol (alpn), domain name and server certificate.

When both the browser and the server support ECH, how much information an attacker can extract from a certificate? 
domain name? time? Get the CA by query up? Or these are encrypted, only password and algorithms?

**Reply by mydogshitgold** [reply]
> When both the browser and the server support ECH, how much information an attacker can extract from a certificate?
domain name? time? Get the CA by query up? Or these are encrypted, only password and algorithms?

If I recall correctly, the ClientHelloInner is completely encrypted, so an attacker in the privileged position can only extract TLS version, packet length, timing, ECH public key, cipher suite, and additionally the whole (not so important) ClientHelloOuter. I don't think they can extract the actual SNI. But you should know that the attacker may just try to force a clear text client hello by dropping or blocking any ECH packets, and they have every reason to do so.

If you want to see for yourself you can capture some QUIC initial packets with wireshark, whichs support decrypting them. Currently the [ECH private key used in QUIC is well-known](https://datatracker.ietf.org/doc/html/draft-ietf-quic-tls-33#section-5.2).

---

You can read this [blog by cloudflare about ECH](https://blog.cloudflare.com/encrypted-client-hello/)

**Reply by nursery01** [reply]
> If I recall correctly, the ClientHelloInner is completely encrypted, so an attacker in the privileged position can only extract TLS version, packet length, timing, ECH public key, cipher suite, and additionally the whole (not so important) ClientHelloOuter. I don't think they can extract the actual SNI. But you should know that the attacker may just try to force a clear text client hello by dropping or blocking any ECH packets, and they have every reason to do so.

I opened ECH in firefox. ->[https://github.com/nursery01/browser-ech](https://github.com/nursery01/browser-ech)
I tried fetching data before, but i see "crypto.cloudflare.com" in client hello. I don`t know why. 

> If you want to see for yourself you can capture some QUIC initial packets with wireshark, whichs support decrypting them. Currently the [ECH private key used in QUIC is well-known](https://datatracker.ietf.org/doc/html/draft-ietf-quic-tls-33#section-5.2).

I was just about to check server hello, my pc power is cut✂️by family, so experiment end.
<br><br>
Maybe my way is wrong😐

**Reply by mydogshitgold** [reply]
> I tried fetching data before, but i see "crypto.cloudflare.com" in client hello. I don`t know why

That's the ClientHelloOuter, and `crypto.cloudflare.com` is the client-facing server for cloudflare edge. If you are confused, read the [IETF draft](https://www.ietf.org/archive/id/draft-ietf-tls-esni-16.html).

**Reply by nursery01** [reply]
> That's the ClientHelloOuter, and is the client-facing server for cloudflare edge. If you are confused, read the [IETF draft](https://www.ietf.org/archive/id/draft-ietf-tls-esni-16.html).`crypto.cloudflare.com`

ok, i understood now. But the browser still exposes the sni

Maybe v2ray can exploit this mechanism to cheat
```
ClientHelloOuter sni=www.cloudflare.com (false)
ClientHelloInner sni=cute.cat.abc (true)
```

![img](https://blog.cloudflare.com/content/images/2020/12/image3-17.png)
